### PR TITLE
Make directed callgraph include nodes with no edges

### PR DESCRIPTION
**Pull request by owen-mc-diffblue**

### CODE REVIEW

**Comment by tautschnig** [comment]
Would you mind explaining (in the commit message) _why_ this is a desirable feature?

File: src/analyses/call_graph.cpp [hunk 1]
```diff
@@ -192,11 +192,17 @@ class function_indicest
...
+call_grapht::directed_grapht
+call_grapht::get_directed_graph(const goto_modelt &goto_model) const
 {
   call_grapht::directed_grapht ret;
   function_indicest function_indices(ret);
 
+  // To make sure we include unreachable functions we first create indices
+  // for all functions in the function map
+  for(const auto &f : goto_model.goto_functions.function_map)
+    function_indices[f.first];
```
Comment (tautschnig) [inline src/analyses/call_graph.cpp]: This function is only computing a different representation of the same underlying graph. Do your changes just fix up a bit of information that is present in  the original call graph, or will it actually yield a different graph?

**Comment by owen-mc-diffblue** [comment]
@tautschnig Thanks for your review. Thinking about it further I realise this is the wrong fix, because it assumes the callgraph has been created for all goto functions.

I still think there might be an underlying problem: `callgrapht` only stores edges, so it has no way of storing a node without any edges. So if you create the callgraph of all functions and then get the directed graph from it then it will be missing nodes that you would expect to be there. Perhaps this is the intended behaviour of `callgrapht`, and the error is in my assumption that getting a directed graph from it would give me a graph with nodes corresponding to the functions that were passed to the constructor of the callgraph. I would value your opinion on this. If I am correct then the fix would be to keep a list of nodes when callgraph is created and use that in `get_directed_graph()`.

**Comment by tautschnig** [comment]
I'd say that it's a _call_ graph, hence nodes without edges don't seem to fit the bill as there's no call to be found. But then you expected to find those here, so you will have different requirements. If you can share your requirements, maybe that helps to come up with an answer? From what I've seen so far I'd say you need both the call graph and the list of functions (which can independently be obtained).

**Comment by smowton** [comment]
Personally I'd expect "the call graph of {f, g, h}" to yield a graph containing exactly three nodes, and edges representing calls, not for functions to disappear entirely if they don't have callers or callees.

One reason for this: it is expected that a call graph would not always be totally connected, but for us to retain both independent subgraphs in that case. In the particular case where one of the components was  of size 1 (i.e. a single node with no associated edges), I don't think that special case should be treated differently.

(Example: if there are edges a -> b and c -> d, we would retain all 4 nodes, but graphs {a, b} and {c, d} are unconnected. It's surprising to me if in the case a -> b, a -> c we don't create graphs {a, b, c} and {d}, but rather throw the 'd' graph away entirely)

Long story short I suggest @owen-jones-diffblue should make the original `call_grapht` retain a vector of functions it was originally constructed upon, to be used if/when `get_directed_graph` is called.

**Comment by owen-mc-diffblue** [comment]
I agree with @smowton, and I have implemented his suggestion.

**Comment by tautschnig** [comment]
Can you please confirm that `goto-instrument --reachable-call-graph` and `goto-instrument --slice-global-inits` are not adversely affected by this change?

File: src/analyses/call_graph.cpp [hunk 2]
```diff
@@ -129,7 +133,9 @@ void call_grapht::add(
   const irep_idt &caller,
   const irep_idt &callee)
 {
-  graph.insert(std::pair<irep_idt, irep_idt>(caller, callee));
+  edges.insert(std::pair<irep_idt, irep_idt>(caller, callee));
```
Comment (NathanJPhillips) [inline src/analyses/call_graph.cpp]: `std::make_pair` or even just brace initialise?

**Review by NathanJPhillips — APPROVED** [review]

**Comment by owen-mc-diffblue** [comment]
@tautschnig They shouldn't be affected because they both use `call_grapht::create_from_root_function()`, so the graph is connected. Also, the graph has at least two nodes because it has `__CPROVER__start` and the user-supplied entry function, so all edges will have an edge going into or out of them. I don't know how to run them but their tests were failing before and are now passing.

**Comment by tautschnig** [comment]
Would you mind doing the extra exercise of adding a regression test for `--reachable-call-graph` (and possibly also `--call-graph`, which now should definitively provide distinct output) in `regression/goto-instrument/`? All it would take is a small C program that has a few functions where not all of them are called from `main`.

**Comment by owen-mc-diffblue** [comment]
@tautschnig I have added those tests. I put them in a commit before the main changes, to signify that they pass before and after the changes.

As you seem to have already noticed, `slice-global-inits-1` already tests that an unreachable function (`func1`) is dealt with appropriately in `goto-instrument --slice-global-inits`.

**Comment by owen-mc-diffblue** [comment]
By the way, this PR should not be merged yet as I have to fix problems caused in a repository which relies on cbmc first.

**Review by tautschnig — APPROVED** [review]

File: src/analyses/call_graph.h [hunk 3]
```diff
@@ -62,13 +62,17 @@ class call_grapht
...
 
-  /// Type of a call graph edge in `grapht`
+  /// Type of the edges in the call graph. Note parallel edges (e.g. A having
+  /// two callsites both targeting B) result in multiple graph edges.
+  typedef std::multimap<irep_idt, irep_idt> edgest;
+
+  /// Type of a call graph edge in `edgest`
   typedef std::pair<irep_idt, irep_idt> edget;
 
+
```
Comment (smowton) [inline src/analyses/call_graph.h]: don't add blank line

**Review by smowton — APPROVED** [review]
Looks great, tiny nitpick

**Comment by owen-mc-diffblue** [comment]
https://github.com/diffblue/test-gen/pull/1717
